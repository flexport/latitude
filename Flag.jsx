/**
 * TEAM: frontend_infra
 * WATCHERS: notandrewkaye
 * @flow
 */

import * as React from "react";
import {css, StyleSheet} from "aphrodite";

import typeof {countries as countriesType} from "./constants/AddressOptionConstants.generated";

type Country = $Keys<countriesType>;
const PATH = "https://assets.flexport.com/flags/svg/1/";
const EXTENSION = ".svg";

type Props = {|
  /** The `ISO 3166-1 aplpha` 2 character country code that you wish to display */
  +countryCode: Country,
  /** Flags have square dimensions and will default to width: 100% unless a maxWidth is specified. */
  +maxWidth?: number,
|};

type State = {|
  loaded: boolean,
  error?: any,
|};

/**
 * @short An SVG icon component specifically for country flags.
 * @brandStatus V2
 * @status Stable
 * @category General
 * @extends React.Component */
export default class Flag extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      loaded: false,
    };
  }
  handleLoaded = () => {
    this.setState({loaded: true});
  };
  render() {
    const {countryCode, maxWidth} = this.props;
    return (
      <img
        src={`${PATH}${countryCode}${EXTENSION}`}
        alt={`${countryCode} Flag`}
        className={css(styles.fade, this.state.loaded && styles.fadeIn)}
        style={{
          width: "100%",
          maxWidth: maxWidth ? `${maxWidth}px` : null,
          height: maxWidth ? `${maxWidth}px` : "auto",
        }}
        onLoad={this.handleLoaded}
      />
    );
  }
}

const styles = StyleSheet.create({
  fade: {
    opacity: 0,
    transition: "opacity 0.2s cubic-bezier(.42,0,.58,1)",
  },
  fadeIn: {
    opacity: 1,
  },
});
