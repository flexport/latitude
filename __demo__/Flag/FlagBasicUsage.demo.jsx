/**
 * TEAM: frontend_infra
 * @flow
 */

import React from "react";
import Flag from "../../Flag";

/**
 * @title Basic usage
 */
export default function FlagBasicUsage() {
  return <Flag countryCode="AU" maxWidth={64} />;
}
