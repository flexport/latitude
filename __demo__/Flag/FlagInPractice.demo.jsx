/**
 * TEAM: frontend_infra
 * @flow
 */

import React from "react";
import Flag from "../../Flag";
import Text from "../../Text";
import DeprecatedVerticalGroup from "../../DeprecatedVerticalGroup";

/**
 * @title In-practice example
 */
export default function FlagInPractice() {
  return (
    <DeprecatedVerticalGroup>
      <Flag countryCode="CA" maxWidth={32} />
      <Text weight="bold" scale="title">
        Canada
      </Text>
      <Text>Pretty great, eh?</Text>
    </DeprecatedVerticalGroup>
  );
}
