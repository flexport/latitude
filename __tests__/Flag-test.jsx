/**
 * TEAM: frontend_infra
 * WATCHERS: notandrewkaye
 * @flow
 */
import {mount, shallow} from "enzyme";
import * as React from "react";

import Flag from "../Flag";

function shallowFlag(propOverrides: {} = {}) {
  const defaultProps = {
    countryCode: "US",
  };

  // $FlowUpgradeFixMe(0.110.1 -> 0.111.1)
  const mergedProps = {...defaultProps, ...propOverrides};
  return shallow(<Flag {...mergedProps} />);
}

describe("Flag", () => {
  describe("can render a flag", () => {
    it("works as expected", () => {
      expect(mount(<Flag countryCode="AU" />)).toMatchSnapshot();
    });
    it("can set a maxHeight", () => {
      const component = shallowFlag({countryCode: "CA", maxWidth: 64});
      expect(component.find("img").props().style).toEqual({
        height: "64px",
        maxWidth: "64px",
        width: "100%",
      });
      expect(component).toMatchSnapshot();
    });
  });
});
