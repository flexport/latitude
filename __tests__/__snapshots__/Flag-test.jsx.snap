// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flag can render a flag can set a maxHeight 1`] = `
<img
  alt="CA Flag"
  className="fade_1gq9q0"
  onLoad={[Function]}
  src="https://assets.flexport.com/flags/svg/1/CA.svg"
  style={
    Object {
      "height": "64px",
      "maxWidth": "64px",
      "width": "100%",
    }
  }
/>
`;

exports[`Flag can render a flag works as expected 1`] = `
<Flag
  countryCode="AU"
>
  <img
    alt="AU Flag"
    className="fade_1gq9q0"
    onLoad={[Function]}
    src="https://assets.flexport.com/flags/svg/1/AU.svg"
    style={
      Object {
        "height": "auto",
        "maxWidth": null,
        "width": "100%",
      }
    }
  />
</Flag>
`;
